tests/org.eclipse.ui.tests.harness/src/org/eclipse/ui/tests/harness/util/RCPTestWorkbenchAdvisor.java

```diff
@@ -199,6 +212,38 @@ public void run() {
 	@Override
 	public void postStartup() {
 		super.postStartup();
+
+		// Wait for all async/sync operations to be scheduled (not blocking UI thread)
+		try {
+			// Wait up to 5 seconds for all operations to be scheduled
+			boolean completed = asyncLatch.await(5, TimeUnit.SECONDS);
+			if (!completed) {
+				throw new AssertionError("Not all async/sync operations were scheduled within timeout");
+			}
+		} catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new RuntimeException("Interrupted while waiting for async/sync operations", e);
+		}
+
+		// Pump the event loop to ensure async runnables execute before marking as started
+		// This prevents the original race condition where async variables might not be set yet
+		Display display = Display.getCurrent();
+		if (display != null) {
+			// Process pending async events multiple times to ensure they execute
+			for (int i = 0; i < 10; i++) {
+				while (display.readAndDispatch()) {
+					// Keep processing events
+				}
+				// Small yield to allow any final async operations to complete
+				try {
+					Thread.sleep(10);
+				} catch (InterruptedException e) {
+					Thread.currentThread().interrupt();
+					break;
+				}
+			}
+		}
```

Comment:
Is this quite complex event processing really necessary? All relevant runnables are guaranteed to be scheduled here, so `display.readAndDispatch()` should not yield `false` until they all have been processed. You just need to make sure that those runnables that are _not_ supposed to be executed during startup had the chance to accidentally be executed (to detect regression). So wouldn't be something like this be sufficient?
```suggestion
		UITestUtil.processEventsUntil(() -> syncWithDisplayAccess && asyncWithDisplayAccess, 5000);
		UITestUtil.processEvents();
```